rfcs/2020-08-27-3603-postgres-metrics.md

```diff
@@ -0,0 +1,154 @@
+- `pg_stat_database_blks_read` tagged with db, host, server, user (counter)
+- `pg_stat_database_stats_reset` tagged with db, host, server, user(counter)
+- `pg_stat_bgwriter_buffers_alloc` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_backend` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_backend_fsync` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_checkpoint` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_buffers_clean` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoint_sync_time` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoint_write_time` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoints_req` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_checkpoints_time` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_maxwritten_clean` tagged with db, host, server, user (counter)
+- `pg_stat_bgwriter_stats_reset` (counter)
+- `pg_stat_database_conflicts` tagged with db, host, server, user (counter)
+- `pg_stat_database_datid` tagged with db, host, server, user (counter)
+- `pg_stat_database_deadlocks` tagged with db, host, server, user (counter)
+- `pg_stat_database_numbackends` tagged with db, host, server, user (gauge)
+- `pg_stat_database_temp_bytes` tagged with db, host, server, user (counter)
+- `pg_stat_database_temp_files` tagged with db, host, server, user (counter)
+- `pg_stat_database_tup_deleted` tagged with db, host, server, user (counter)
+- `pg_stat_database_tup_fetched` tagged with db, host, server, user (counter)
+- `pg_stat_database_tup_inserted` tagged with db, host, server, user (counter)
+- `pg_stat_database_tup_returned` tagged with db, host, server, user (counter)
+- `pg_stat_database_tup_updated` tagged with db, host, server, user (counter)
+- `pg_stat_database_xact_commit` tagged with db, host, server, user (counter)
+- `pg_stat_database_xact_rollback` tagged with db, host, server, user (counter)
+- `postgresql_database_conflicts` tagged by db name and server (counter)
+- `postgresql_database_conflicts_confl_bufferpin` tagged by db name and server (counter)
+- `postgresql_database_conflicts_confl_deadlock` tagged by db name and server (counter)
+- `postgresql_database_conflicts_confl_lock tagged` by db name and server (counter)
+- `postgresql_database_conflicts_confl_snapshot` tagged by db name and server (counter)
+- `postgresql_database_conflicts_confl_tablespace` tagged by db name and server (counter)
+
+Naming of metrics is determined via:
+
+- `pg _ db_name _ column_name`
+
+This is in line with the Prometheus naming convention for their exporter.
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+[sources.my_source_id]
+  type = "postgresql_metrics" # required
+  address = "postgres://postgres@localhost" # required - address of the PG server.
```

Comment:
That's a bit messy as a UI. :( And adding the DB on the URL only solves for one DB per connection, what if I want two?

Reply:
Yeah, it does feel messier. You could just specify the URL twice if you wanted two databases (or maybe do commas?), but I can see this would be weird for the common case of just wanting to monitor all databases on a given server. To that end, we should probably make it easy to just monitor all database (or provide a list of exclusions even). I think it's probably pretty common that I'd want to monitor all of them aside from the `template*` ones.

I think we should come up with a rule of thumb for when we should do single vs. multiple endpoints for a given source though; to ease the decision in the future.

Reply:
Hmmm I've always assumed you'd specific the source twice for separate servers... Specifying the connection URL twice will double (or more for other DBs) the connections and mean multiple collections on a single DB server. I don't think that's a good performance experience for folks.

Reply:
Yeah, I definitely agree in this case.

I'm wondering what you think about the `apache_metrics` case though. I did change it to `endpoints` in response to https://github.com/timberio/vector/pull/3519#discussion_r475200386 . The prometheus scraper also allows multiple `endpoints`.

Reply:
I feel like Apache and PG are differently architected and hence monitored differently. Apache is a single server you're monitoring. PG is monitoring one or more/all databases in a PG server. That additional level of granularity makes `endpoints` odd to me.

Reply:
I've reverted this to `endpoint` for now. We can discuss.